### PR TITLE
JBPM-8731 Default runtime strategy of 'New Deployment Unit' wizard s…

### PR DESCRIPTION
**Pull request by bxf12315**
…hould respect kie-deployment-descriptor.xml

Child PR: https://github.com/kiegroup/kie-wb-common/pull/2654

### CODE REVIEW

**Comment by bxf12315** [comment]
@cristianonicolai please review it

**Comment by ederign** [comment]
@cristianonicolai hi buddy, who is the best QE for this?

**Comment by cristianonicolai** [comment]
@ederign that would be @jstastny-cz

**Comment by jstastny-cz** [comment]
Hello, the feature is not planned for current version. I won't approve the PR until there are distinct branches.

**Comment by ederign** [comment]
@bxf12315 hi, next time could you please add the child PR's to the description? Thanks.

**Comment by bxf12315** [comment]
> @bxf12315 hi, next time could you please add the child PR's to the description? Thanks.

ok Thanks

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/M2RepoServiceImpl.java [hunk 1]
```diff
@@ -260,6 +260,18 @@ public String getRepositoryURL() {
         }
     }
 
+    @Override
+    public String getKMoudleText(String path) {
```
Comment (cristianonicolai) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/M2RepoServiceImpl.java]: @bxf12315 would be good to add tests for these methods as well.

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java [hunk 2]
```diff
@@ -336,4 +337,12 @@ public void testGetPomTextVerifiesPath() {
         exception.expect(RuntimeException.class);
         repo.getPomText("dir/name.foo");
     }
+
+    @Test
+    public void testLoadFileTextFromJar() {
+        File jarFile = new File("src/test/resources/evaluation-12.1.1.jar");
+
+        assertNotNull(GuvnorM2Repository.loadFileTextFromJar(jarFile,"META-INF","kie-deployment-descriptor.xml"));
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java]: also, please test some failure scenarios like invalid path, null parameters, etc.

**Comment by bxf12315** [comment]
@cristianonicolai @nmirasch updated.Please review again.

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java [hunk 3]
```diff
@@ -460,12 +460,35 @@ public String getPomText(final String path) {
         }
     }
 
+    public String getKMoudleText(final String path) {
```
Comment (jstastny-cz) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: KMoudle -> KModule

Comment (bxf12315) [reply uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: updated

**Comment by bxf12315** [comment]
jenkins please retest

**Comment by bxf12315** [comment]
jenkins please retest

**Comment by bxf12315** [comment]
jenkins please retest

**Comment by bxf12315** [comment]
Jenkins execute full downstream build

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/M2RepoServiceImplM2RepoURLTest.java [hunk 4]
```diff
@@ -90,4 +94,17 @@ public void repoURLFromSystemPropertyInvalidURL() throws Exception {
         assertEquals("file://path-to-m2", m2RepoService.getRepositoryURL());
         verify(logger).warn(anyString());
     }
-}
\ No newline at end of file
+
+    @Test
+    public void testGetKMoudleAndKieDeploymentDescriptorText() throws Exception {
```
Comment (jstastny-cz) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/M2RepoServiceImplM2RepoURLTest.java]: Moudle->Module

File: uberfire-m2repo-editor/uberfire-m2repo-editor-api/src/main/java/org/guvnor/m2repo/service/M2RepoService.java [hunk 5]
```diff
@@ -51,4 +51,18 @@
      * @return A String representing the repository's URL relative to the container's root
      */
     String getRepositoryURL();
+
+    /**
+     * Retrieve the kmodule.xml from the given Path.
+     * @param path The path to the file
+     * @return The kmodule.xml text
+     */
+    String getKMoudleText(String path);
```
[the remark this answers is not in the document]
Comment (bxf12315) [reply uberfire-m2repo-editor/uberfire-m2repo-editor-api/src/main/java/org/guvnor/m2repo/service/M2RepoService.java]: It is old version.

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/M2RepoServiceImplM2RepoURLTest.java (hunk 4 above)
Comment (bxf12315) [reply uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/M2RepoServiceImplM2RepoURLTest.java]: updated

**Comment by bxf12315** [comment]
Jenkins execute full downstream build

**Comment by bxf12315** [comment]
Jenkins execute full downstream build

**Comment by bxf12315** [comment]
jenkins please retest

**Comment by bxf12315** [comment]
Jenkins execute full downstream build

**Comment by bxf12315** [comment]
Jenkins execute full downstream build

**Comment by nmirasch** [comment]
Jenkins execute full downstream build

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java [hunk 6]
```diff
@@ -27,6 +27,7 @@
 import java.util.List;
 import javax.enterprise.inject.Instance;
 
+import org.apache.commons.io.FileUtils;
```
Comment (nmirasch) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java]: @bxf12315  Unused import here

Comment (bxf12315) [reply uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java]: Done

**Comment by cristianonicolai** [comment]
@jstastny-cz can you please review again, thanks

**Comment by bxf12315** [comment]
@jstastny-cz @cristianonicolai @nmirasch Please review it. If it's ok, I will rebase it.

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java [hunk 7]
```diff
@@ -336,4 +338,23 @@ public void testGetPomTextVerifiesPath() {
...
+
+        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, "META-INF", "kie-descriptor.xml"));
+        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, "META-INF", "modu.xml"));
+
+        assertEquals("", GuvnorM2Repository.loadFileTextFromJar(jarFile, "META-INF", ""));
+        assertEquals("", GuvnorM2Repository.loadFileTextFromJar(jarFile, "META-INF", ""));
+
+        exception.expect(NullPointerException.class);
+        assertNull("", GuvnorM2Repository.loadFileTextFromJar(jarFile, "META-INF", null));
+        exception.expect(NullPointerException.class);
```
Comment (jstastny-cz) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java]: expecting NPE is not acceptable, improve the tested method.

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java [hunk 8]
```diff
@@ -658,35 +677,35 @@ public String generatePOM(final GAV gav) {
...
+        sBuilder.append(GAV.VERSION_NAME + "version=");
         sBuilder.append(gav.getVersion());
         sBuilder.append("\n");
 
         return sBuilder.toString();
     }
 
     public String getPomXmlPath(final GAV gav) {
-        return "META-INF/maven/" + gav.getGroupId() + "/" + gav.getArtifactId() + "/pom.xml";
+        return META_INF_MAVEN+"/" + gav.getGroupId() + "/" + gav.getArtifactId() + "/"+POM_XML;
```
Comment (jstastny-cz) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: formatting of the code

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java [hunk 9]
```diff
@@ -658,35 +677,35 @@ public String generatePOM(final GAV gav) {
...
     }
 
     public String getPomPropertiesPath(final GAV gav) {
-        return "META-INF/maven/" + gav.getGroupId() + "/" + gav.getArtifactId() + "/pom.properties";
+        return META_INF_MAVEN+"/" + gav.getGroupId() + "/" + gav.getArtifactId() + "/" + POM_PROPERTIES;
```
Comment (jstastny-cz) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: formatting of the code

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java [hunk 10]
```diff
@@ -336,4 +338,23 @@ public void testGetPomTextVerifiesPath() {
         exception.expect(RuntimeException.class);
         repo.getPomText("dir/name.foo");
     }
+
+    @Test
+    public void testLoadKieTextFromJar() {
```
[the remark this answers is not in the document]
Comment (bxf12315) [reply uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java]: 1 here's file not common file, it's include kie-deployment-descriptor.xml and kmodule.xml, so I think kie text is better.

2 it just test kie text not test loadKieTextFromJar, the [case](https://github.com/kiegroup/appformer/pull/706/files#diff-a1b139ce9e897c2b9719bdb8490c9f56R637) for loadKieTextFromJar.

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/M2MavenRepositoryServiceImplTest.java [hunk 11]
```diff
@@ -632,4 +632,26 @@ private void deployArtifact(GAV gav) {
...
+    public void testGetKieText() {
+        GAV gavEvaluation = new GAV("evaluation", "evaluation", "12.1.1-SNAPSHOT");
+        InputStream is = this.getClass().getResourceAsStream("evaluation-12.1.1.jar");
+
+        repo.deployArtifact(is, gavEvaluation, false);
+
+        Collection<File> files = repo.listFiles();
+        for (File file : files) {
+            String fileName = file.getName();
+            if (fileName.startsWith("evaluation-12.1.1") && fileName.endsWith(".jar")) {
```
[the remark this answers is not in the document]
Comment (bxf12315) [reply uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/M2MavenRepositoryServiceImplTest.java]: I'm not sure what you means. Please give me more details.

**Comment by bxf12315** [comment]
@jstastny-cz I add some test, please check it.

File: uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/model/GAV.java [hunk 12]
```diff
@@ -26,6 +26,10 @@
 @Portable
 public class GAV implements Serializable {
 
+    public static final String GROUP_ID_NAME = "groupId";
```
[the remark this answers is not in the document]
Comment (bxf12315) [reply uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/model/GAV.java]: first, these constants will use by GuvnorM2Repository.
second, it is smells issue of sonar, so I refator them to constants.

[the remark this answers is not in the document]
Comment (cristianonicolai) [reply uberfire-project/uberfire-project-api/src/main/java/org/guvnor/common/services/project/model/GAV.java]: this is actually my suggestion, to move these constants to GuvnorM2Repository.

**Comment by bxf12315** [comment]
@cristianonicolai @jstastny-cz updated

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java [hunk 13]
```diff
@@ -460,22 +464,47 @@ public String getPomText(final String path) {
...
+        }
+        return "";
+    }
+
+    private static String loadPomFromJar(final File file) {
+        return loadFileTextFromJar(file, META_INF_MAVEN, POM_XML);
+    }
+
+    protected static String loadFileTextFromJar(final File jarFile, String path, String fileName) {
+        if (jarFile != null && !"".equals(fileName) && fileName != null) {
```
Comment (jstastny-cz) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: jarFile != null &&  fileName != null && !fileName.isEmpty() is more readable IMO.

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java [hunk 14]
```diff
@@ -460,22 +464,47 @@ public String getPomText(final String path) {
         }
     }
 
-    private static String loadPomFromJar(final File file) {
-        try (ZipFile zip = new ZipFile(file)) {
-            for (Enumeration e = zip.entries(); e.hasMoreElements(); ) {
-                ZipEntry entry = (ZipEntry) e.nextElement();
+    public String getKModuleText(final String path) {
+        return loadKieTextFromJar(path, META_INF, "kmodule.xml");
```
Comment (jstastny-cz) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: to be consistent we should introduce constant for kmodule.xml

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java [hunk 15]
```diff
@@ -460,22 +464,47 @@ public String getPomText(final String path) {
...
+    }
 
-                if (entry.getName().startsWith("META-INF/maven") && entry.getName().endsWith("pom.xml")) {
-                    return zipEntryToString(zip,
-                                            entry);
+    public String getKieDeploymentDescriptorText(final String path) {
+        return loadKieTextFromJar(path, META_INF, "kie-deployment-descriptor.xml");
```
Comment (jstastny-cz) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: to be consistent we should introduce constant for kie-deployment-descriptor.xml and possibly also use it in the tests

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java [hunk 16]
```diff
@@ -460,22 +464,47 @@ public String getPomText(final String path) {
...
+        ArtifactRepository repository = this.getArtifactRepository(ArtifactRepositoryService.GLOBAL_M2_REPO_NAME);
+        final File file = new File(repository.getRootDir(),
+                                   jarFilePath);
+
+
+        final String normalizedPath = file.toPath().normalize().toString();
+        if (isJar(normalizedPath) || isKJar(normalizedPath)) {
+            return loadFileTextFromJar(file, subPath, fileName);
+        }
+        return "";
```
Comment (jstastny-cz) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/GuvnorM2Repository.java]: TLDR; I would return null here.
I can see a two ways we're handling "issue with the file" .... here, we're returning empty string in case the filename does NOT end with .jar or .kjar.
If it does end with the suffixes, but the file is corrupt or error during reading happens, we return null from loadFileTextFromJar method.
Empty string should be IMO returned only when the file exists, is a jar/kjar has a corresponding zipentry and the entry is an empty file.

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/M2MavenRepositoryServiceImplTest.java [hunk 17]
```diff
@@ -632,4 +636,24 @@ private void deployArtifact(GAV gav) {
...
     }
+
+    @Test
+    public void testGetKieText() {
+        GAV gavEvaluation = new GAV("evaluation", "evaluation", "12.1.1-SNAPSHOT");
+        InputStream is = this.getClass().getResourceAsStream("evaluation-12.1.1.jar");
+
+        repo.deployArtifact(is, gavEvaluation, false);
+
+        Collection<File> files = repo.listFiles();
```
Comment (cristianonicolai) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/M2MavenRepositoryServiceImplTest.java]: this line should be removed

**Comment by jstastny-cz** [comment]
jenkins please retest this

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java [hunk 18]
```diff
@@ -336,4 +338,20 @@ public void testGetPomTextVerifiesPath() {
...
+        assertNotNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, GuvnorM2Repository.KIE_DEPLOYMENT_DESCRIPTOR_XML));
+        assertNotNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, GuvnorM2Repository.KMODULE_XML));
+
+        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, "kie-descriptor.xml"));
+        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, "modu.xml"));
+
+        assertEquals(null, GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, ""));
+        assertEquals(null, GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, ""));
+
+        assertNull(null, GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, null));
```
Comment (cristianonicolai) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java]:
```suggestion
        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, null));
```

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java [hunk 19]
```diff
@@ -336,4 +338,20 @@ public void testGetPomTextVerifiesPath() {
...
+        assertNull(null, GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, null));
```
Comment (cristianonicolai) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java]:
```suggestion
        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, null));
```

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java [hunk 20]
```diff
@@ -336,4 +338,20 @@ public void testGetPomTextVerifiesPath() {
...
+    @Test
+    public void testLoadFileTextFromJar() {
+        File jarFile = new File("src/test/resources/org/guvnor/m2repo/backend/server/evaluation-12.1.1.jar");
+        assertNotNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, GuvnorM2Repository.KIE_DEPLOYMENT_DESCRIPTOR_XML));
+        assertNotNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, GuvnorM2Repository.KMODULE_XML));
+
+        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, "kie-descriptor.xml"));
+        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, "modu.xml"));
+
+        assertEquals(null, GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, ""));
```
Comment (cristianonicolai) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java]:
```suggestion
        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, ""));
```

File: uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java [hunk 21]
```diff
@@ -336,4 +338,20 @@ public void testGetPomTextVerifiesPath() {
...
+        assertEquals(null, GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, ""));
```
Comment (cristianonicolai) [inline uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/test/java/org/guvnor/m2repo/backend/server/GuvnorM2RepositoryTest.java]:
```suggestion
        assertNull(GuvnorM2Repository.loadFileTextFromJar(jarFile, GuvnorM2Repository.META_INF, ""));
```

**Comment by bxf12315** [comment]
@cristianonicolai updated

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by cristianonicolai** [comment]
@bxf12315 this PR probably needs rebase

**Comment by bxf12315** [comment]
@cristianonicolai done

**Comment by cristianonicolai** [comment]
@ederign, please merge this PR along with https://github.com/kiegroup/kie-wb-common/pull/2654 thanks.

**Comment by cristianonicolai** [comment]
@ederign @porcelli please merge this PR along with kiegroup/kie-wb-common#2654 thanks.

**Comment by ederign** [comment]
@cristianonicolai sorry for the delay.
